text/3573-is.md

```diff
@@ -0,0 +1,284 @@
+<https://doc.rust-lang.org/reference/patterns.html> for details on patterns.
+
+If the `EXPR` matches the `PATTERN`, the `is` expression evaluates to `true`,
+and additionally binds any bindings specified in `PATTERN` in the current scope
+for code subsequently executed along the path where the `is` expression
+is known to be `true`.
+
+For example:
+
+```rust
+if an_option is Some(x) && x > 3 {
+    println!("{x}");
+}
+```
+
+The bindings in the pattern are not bound along any code path potentially
+reachable where the expression did not match:
+
+```rust
+if (an_option is Some(x) && x > 3) || (more_conditions /* x is not bound here*/) {
+    // x is not bound here
+} else {
+    // x is not bound here
+}
+// x is not bound here
+```
+
+The pattern may use alternation (within parentheses), but must have the same
+bindings in every alternative:
+
+```rust
+if color is (RGB(r, g, b) | RGBA(r, g, b, _)) && r == b && g < 10 {
+    println!("condition met")
+}
+
+// ERROR: `a` is not bound in all alternatives of the pattern
+if color is (RGB(r, g, b) | RGBA(r, g, b, a)) && r == b && g < 10 {
+    println!("condition met")
+}
+```
+
+`is` may appear anywhere a boolean expression is accepted:
+
+```rust
+func(x is Some(y) && y > 3);
+```
+
+The `is` operator may not appear as a statement; use `let` to bind a pattern in
+that context:
+
+```rust
+// ERROR: use `let` for this
+an_expression() is x;
+
+let x = an_expression();
+```
+
+# Reference-level explanation
+
+Add a new [operator
+expression](https://doc.rust-lang.org/reference/expressions/operator-expr.html),
+`IsExpression`:
+
+> **<sup>Syntax</sup>**\
+> _IsExpression_ :\
+> &nbsp;&nbsp; _Expression_ `is` _PatternNoTopAlt_
+
+Add `is` to the [operator
+precedence](https://doc.rust-lang.org/reference/expressions.html#expression-precedence)
+table, at the same precedence level as `==`, and likewise non-associative
+(requiring parentheses).
+
+Detect `is` appearing as a top-level statement and produce an error, with a
+rustfix suggestion to use `let` instead.
+
+# Drawbacks
+
+Introducing both the `is` operator and `let`-chains would provide two different
+ways to do a pattern match as part of a condition. Having more than one way to
+do something could lead people to wonder if there's a difference; we would need
+to clearly communicate that they serve similar purposes.
+
+An `is` operator will produce a name conflict with [the `is` method on
+`dyn Any`](https://doc.rust-lang.org/std/any/trait.Any.html#method.is) in the
+standard library, and with the (relatively few) methods named `is` in the
+ecosystem. This will not break any existing Rust code, as the operator will
+only exist in the Rust 2024 edition and newer. The Rust standard library and
+any other library that wants to avoid requiring the use of `r#is` in Rust 2024
+and newer could provide aliases of these methods under a new name; for
+instance, the standard library could additionally provide `Any::is` under a new
+name `is_type`.
```

Comment:
I personally would just promote adding this alias to the body of the RFC, including deprecating the old name. It seems like a clear-cut solution to the problem and while deprecating a method would be a drawback, I think that it should be tracked as part of the RFC and done well in advance of the feature being implemented to ensure people have time to migrate.

Reply:
@clarfonthey 1) I don't think we should deprecate the old name. 2) I don't want to combine the library proposal (and bikeshedding on the name) with this language proposal. 3) If this is accepted, I'll immediately propose a PR for the library and link that from the tracking issue.

Reply:
I added an item to the unresolved questions section to ensure that this gets tracked in the tracking issue.